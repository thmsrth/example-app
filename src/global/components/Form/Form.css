:root {
  /** Typography **/
  --form-input-label-font-size: 1.5rem;
  --form-input-label-font-weight: 600;

  --form-input-alert-message-font-size: 14px;
  --form-input-alert-message-font-weight: 400;
  --form-input-alert-message-font-style: italic;
  --form-input-alert-message-color: var(--color-alert);

  --form-input-warning-message-font-size: 15px;
  --form-input-warning-message-font-weight: 400;
  --form-input-warning-message-font-style: italic;
  --form-input-warning-message-color: var(--color-warning);
}

.Form__form {
  margin: 0 auto;
  padding: var(--section-padding);
  border: 1px solid var(--border-color);
}

.Form__icon {
  margin: 0 var(--section-margin);
}

.Form__cancel {
  background-color: var(--action-cancel-background-color);
  color: var(--action-cancel-color);
}

.Form__submit {
  background-color: var(--action-submit-background-color);
  color: var(--action-submit-color);
}

.Form__submit__create,
.Form__submit__add {
  background-color: var(--action-create-background-color);
  color: var(--action-create-color);
}

.Form__fields {
  padding: var(--section-padding);
  display: block;
}

.FormInput__label {
  display: block;
  padding: var(--section-padding);
  font-size: var(--form-input-label-font-size);
  font-weight: var(--form-input-label-font-weight);
  width: 100%;
  margin-bottom: 0;
}

.FormInput__label--disabled {
  color: var(--input-disabled-color);
}

.FormInput__alertMsg {
  display: none;
  font-size: var(--form-input-alert-message-font-size);
  font-weight: var(--form-input-alert-message-font-weight);
  font-style: var(--form-input-alert-message-font-style);
  margin-top: calc(var(--section-margin) / 2);
}

.FormInput--alert > .FormInput__label > .FormInput__input,
.FormInput--alert > .FormInput__label > .InputCollection__fields {
  border-color: var(--form-input-alert-message-color);
  margin-top: 0;
}

.FormInput__alertMsg {
  color: var(--form-input-alert-message-color);
  display: block;
}

.FormInput__warningMsg {
  font-size: var(--form-input-warning-message-font-size);
  font-weight: var(--form-input-warning-message-font-weight);
  font-style: var(--form-input-warning-message-font-style);
  color: var(--form-input-warning-message-color);
  display: block;
}

.FormInput__input {
  margin: calc(var(--section-margin) / 2) 0;
  display: block;
  width: 100%;
}

.FormInput__input--disabled,
.FormInput__input--disabled .Select-value,
.FormInput__input--disabled .Select-placeholder {
  cursor: not-allowed;
  background-color: var(--form-background-color);
}

.FormInput--alert > .FormInput__label > .FormInput__input--disabled {
  border-color: var(--border-color);
}

.Form__formActions {
  padding: var(--section-padding);
  margin-top: var(--section-margin);
  display: flex;
  justify-content: space-evenly;
}

.FormInput__options {
  padding-top: var(--section-padding);
}
