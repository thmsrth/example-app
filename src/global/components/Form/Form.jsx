import React from 'react';

import PropTypes from 'prop-types';
import keysType from './types/keys';

import I18nGlobal from '../../utils/I18nGlobal';
import { bemCls } from '../../utils/ClassNameHelpers';
import { getFormInputField } from '../../utils/FormHelpers';

import Button from '../Button';

import './Form.css';

const Form = ({
  className, keys, data, action, readOnly,
  handleChange, handleBlur, handleSubmit, handleCancel, optionsFromData, loading,
}) => {
  if (data) {
    const isLoading = loading !== '';
    return (
      <form
        name={className}
        action={action}
        id={className}
        onSubmit={handleSubmit}
      >
        <fieldset disabled={isLoading} className={bemCls(`Form ${className}`, '__form')}>
          <div className={bemCls(`Form ${className}`, '__fields')}>
            {keys.map((key) => {
              if (key) {
                const inputData = data[key.name];
                if (inputData !== undefined) {
                  return getFormInputField(
                    key,
                    inputData,
                    className,
                    handleChange,
                    handleBlur,
                    optionsFromData,
                  );
                }
              }
              return null;
            })}
          </div>
          <div className={bemCls(`Form ${className}`, '__formActions')}>
            {!!handleCancel
              && (
                <Button
                  className={bemCls(`Form ${className}`, '__cancel')}
                  iconName="remove"
                  text={I18nGlobal.t('en.actions.cancel')}
                  onClick={handleCancel}
                  disabled={isLoading}
                />
              )
            }
            {!readOnly
              && (
              <Button
                className={`${bemCls(`Form ${className}`, '__submit')} ${
                  bemCls(`Form ${className}`, `__submit__${action}`)}`}
                iconName={action === 'send' ? action : 'check'}
                text={action
                  ? I18nGlobal.t(`en.actions.${action}`)
                  : I18nGlobal.t('en.actions.save')}
                type="submit"
                loading={isLoading}
              />
              )
            }
          </div>
        </fieldset>
      </form>
    );
  }
  return null;
};

export default Form;

Form.propTypes = {
  className: PropTypes.string,
  action: PropTypes.string,
  keys: keysType,
  data: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  // eslint-disable-next-line react/forbid-prop-types
  optionsFromData: PropTypes.object,
  readOnly: PropTypes.bool,
  loading: PropTypes.string,
  handleChange: PropTypes.func,
  handleSubmit: PropTypes.func,
  handleCancel: PropTypes.func,
  handleBlur: PropTypes.func,
};

Form.defaultProps = {
  className: '',
  action: 'create',
  keys: [],
  data: null,
  optionsFromData: null,
  readOnly: false,
  loading: '',
  handleChange: () => {},
  handleSubmit: () => {},
  handleCancel: undefined,
  handleBlur: () => {},
};
