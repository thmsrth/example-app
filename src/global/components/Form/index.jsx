import React, { Component } from 'react';

import PropTypes from 'prop-types';
import keysType from './types/keys';

import {
  getFormData,
  getValidationKeys,
  getValidatedKeyData,
  isEveryKeyDataValid,
  getSubmitDataValues,
} from '../../utils/FormHelpers';

import Form from './Form';

class FormWrapper extends Component {
  constructor(props) {
    super(props);
    this.state = {
      data: getFormData(props.keys, props.data),
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleBlur = this.handleBlur.bind(this);
    this.validateForm = this.validateForm.bind(this);
    this.updateKeysDataValidation = this.updateKeysDataValidation.bind(this);
    this.updateData = this.updateData.bind(this);
  }

  handleChange(keyName, value) {
    const data = this.updateKeyDataFromState(keyName, value);

    const { keys } = this.props;
    const currentKey = keys.find(k => k.name === keyName);

    this.updateData(data);

    if (currentKey && currentKey.onChange) {
      currentKey.onChange(value, data, this.updateData);
    }
  }

  handleBlur(keyName, validationKeys, options = {}) {
    const { value } = options;
    const { data } = this.state;
    const newData = value
      ? this.updateKeyDataFromState(keyName, value)
      : data;

    const validatedData = getValidatedKeyData(keyName, newData, validationKeys, options);

    this.setState({ data: validatedData });
  }

  handleSubmit(event) {
    event.preventDefault();
    const { data } = this.state;
    const { keys } = this.props;
    const isFormValid = this.validateForm();

    if (isFormValid) {
      const { submitParams, handleSubmit, formatData } = this.props;
      const dataValues = getSubmitDataValues(keys, data);

      handleSubmit({ ...submitParams, data: formatData(dataValues) });
    }
  }

  updateData(data) {
    this.setState({ data });
  }

  updateKeyDataFromState(keyName, value) {
    const { data } = this.state;

    return {
      ...data,
      [keyName]: {
        ...data[keyName],
        value,
      },
    };
  }

  validateForm() {
    const data = this.updateKeysDataValidation();
    return isEveryKeyDataValid(data);
  }

  updateKeysDataValidation() {
    const { keys } = this.props;
    const { data } = this.state;

    const validatedData = keys.reduce((d, key) => {
      const valData = getValidatedKeyData(
        key.name,
        data,
        getValidationKeys(key, data),
        { fields: key.fields },
      );

      return {
        ...d,
        [key.name]: valData[key.name],
      };
    }, {});

    this.setState({ data: validatedData });

    return Object.values(validatedData);
  }

  render() {
    const { data } = this.state;
    return (
      <Form
        {...this.props}
        data={data}
        handleChange={this.handleChange}
        handleSubmit={this.handleSubmit}
        handleBlur={this.handleBlur}
      />
    );
  }
}

export default FormWrapper;

FormWrapper.propTypes = {
  action: PropTypes.string,
  // eslint-disable-next-line react/forbid-prop-types
  data: PropTypes.object,
  // eslint-disable-next-line react/forbid-prop-types
  submitParams: PropTypes.object,
  keys: keysType,
  handleSubmit: PropTypes.func,
  formatData: PropTypes.func,
};

FormWrapper.defaultProps = {
  action: '',
  data: null,
  submitParams: null,
  keys: [],
  formatData: data => data,
  handleSubmit: () => {},
};
