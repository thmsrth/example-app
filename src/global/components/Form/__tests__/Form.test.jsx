import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import Form from '../Form';

const data = {
  name: {
    isValid: true,
    showWarning: false,
    value: 'My Name',
  },
  description: {
    isValid: true,
    showWarning: false,
    value: 'My Description',
  },
};

const props = {
  action: 'edit',
  className: 'MyClassForm',
  keys: [
    {
      name: 'name',
      title: 'Name (internal use)*',
      type: 'text',
      format: 'uppercase',
      required: true,
      warningAlert: 'My Warning Message',
      validationAlert: 'My Validation Alert Message',
    },
    {
      name: 'description',
      title: 'Description (internal use)',
      type: 'text',
      format: 'capitalize',
      warningAlert: 'My Warning Message',
      validationAlert: 'My Validation Alert Message',
    },
    null,
  ],
  data,
  handleCancel: () => {},
};

const dataWithWarning = {
  data: {
    name: {
      isValid: true,
      showWarning: true,
      value: 'My Name',
    },
    description: {
      isValid: true,
      showWarning: true,
      value: 'My Description',
    },
  },
};

const dataWithValidationAlert = {
  data: {
    name: {
      isValid: false,
      showWarning: false,
      value: 'My Name',
    },
    description: {
      isValid: false,
      showWarning: false,
      value: 'My Description',
    },
  },
};

describe('Form', () => {
  it('should render the component correctly for the edit action with data', () => {
    const tree = shallow(<Form {...props} />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should render the component correctly for the create action with data', () => {
    const tree = shallow(<Form
      {...props}
      action="create"
    />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should render the component correctly with form/fieldset disabled while loading', () => {
    const tree = shallow(<Form {...props} loading="Loading..." />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should render the component correctly with warning message for each field', () => {
    const tree = shallow(<Form
      {...props}
      {...dataWithWarning}
    />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should render the component correctly with validation alert message for each field', () => {
    const tree = shallow(<Form
      {...props}
      {...dataWithValidationAlert}
    />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});
