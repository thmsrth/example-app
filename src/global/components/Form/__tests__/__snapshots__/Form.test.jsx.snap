// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render the component correctly for the create action with data 1`] = `
<form
  action="create"
  id="MyClassForm"
  name="MyClassForm"
  onSubmit={[Function]}
>
  <fieldset
    className="Form__form MyClassForm__form"
    disabled={false}
  >
    <div
      className="Form__fields MyClassForm__fields"
    >
      <InputTextWrapper
        className="FormInput MyClassFormInput"
        hidden={false}
        key="name"
        label="Name (internal use)*"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        validationKeys={
          Object {
            "onValidate": undefined,
            "onWarning": undefined,
            "pattern": undefined,
            "required": true,
            "validationAlert": "",
            "warningAlert": "",
          }
        }
        value="My Name"
      />
      <InputTextWrapper
        className="FormInput MyClassFormInput"
        hidden={false}
        key="description"
        label="Description (internal use)"
        name="description"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        validationKeys={
          Object {
            "onValidate": undefined,
            "onWarning": undefined,
            "pattern": undefined,
            "required": undefined,
            "validationAlert": "",
            "warningAlert": "",
          }
        }
        value="My Description"
      />
    </div>
    <div
      className="Form__formActions MyClassForm__formActions"
    >
      <ButtonWrapper
        className="Form__cancel MyClassForm__cancel"
        disabled={false}
        iconName="remove"
        onClick={[Function]}
        params={Object {}}
        text="Cancel"
      />
      <ButtonWrapper
        className="Form__submit MyClassForm__submit Form__submit__create MyClassForm__submit__create"
        iconName="check"
        loading={false}
        onClick={[Function]}
        params={Object {}}
        text="Create"
        type="submit"
      />
    </div>
  </fieldset>
</form>
`;

exports[`Form should render the component correctly for the edit action with data 1`] = `
<form
  action="edit"
  id="MyClassForm"
  name="MyClassForm"
  onSubmit={[Function]}
>
  <fieldset
    className="Form__form MyClassForm__form"
    disabled={false}
  >
    <div
      className="Form__fields MyClassForm__fields"
    >
      <InputTextWrapper
        className="FormInput MyClassFormInput"
        hidden={false}
        key="name"
        label="Name (internal use)*"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        validationKeys={
          Object {
            "onValidate": undefined,
            "onWarning": undefined,
            "pattern": undefined,
            "required": true,
            "validationAlert": "",
            "warningAlert": "",
          }
        }
        value="My Name"
      />
      <InputTextWrapper
        className="FormInput MyClassFormInput"
        hidden={false}
        key="description"
        label="Description (internal use)"
        name="description"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        validationKeys={
          Object {
            "onValidate": undefined,
            "onWarning": undefined,
            "pattern": undefined,
            "required": undefined,
            "validationAlert": "",
            "warningAlert": "",
          }
        }
        value="My Description"
      />
    </div>
    <div
      className="Form__formActions MyClassForm__formActions"
    >
      <ButtonWrapper
        className="Form__cancel MyClassForm__cancel"
        disabled={false}
        iconName="remove"
        onClick={[Function]}
        params={Object {}}
        text="Cancel"
      />
      <ButtonWrapper
        className="Form__submit MyClassForm__submit Form__submit__edit MyClassForm__submit__edit"
        iconName="check"
        loading={false}
        onClick={[Function]}
        params={Object {}}
        text="Save"
        type="submit"
      />
    </div>
  </fieldset>
</form>
`;

exports[`Form should render the component correctly with form/fieldset disabled while loading 1`] = `
<form
  action="edit"
  id="MyClassForm"
  name="MyClassForm"
  onSubmit={[Function]}
>
  <fieldset
    className="Form__form MyClassForm__form"
    disabled={true}
  >
    <div
      className="Form__fields MyClassForm__fields"
    >
      <InputTextWrapper
        className="FormInput MyClassFormInput"
        hidden={false}
        key="name"
        label="Name (internal use)*"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        validationKeys={
          Object {
            "onValidate": undefined,
            "onWarning": undefined,
            "pattern": undefined,
            "required": true,
            "validationAlert": "",
            "warningAlert": "",
          }
        }
        value="My Name"
      />
      <InputTextWrapper
        className="FormInput MyClassFormInput"
        hidden={false}
        key="description"
        label="Description (internal use)"
        name="description"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        validationKeys={
          Object {
            "onValidate": undefined,
            "onWarning": undefined,
            "pattern": undefined,
            "required": undefined,
            "validationAlert": "",
            "warningAlert": "",
          }
        }
        value="My Description"
      />
    </div>
    <div
      className="Form__formActions MyClassForm__formActions"
    >
      <ButtonWrapper
        className="Form__cancel MyClassForm__cancel"
        disabled={true}
        iconName="remove"
        onClick={[Function]}
        params={Object {}}
        text="Cancel"
      />
      <ButtonWrapper
        className="Form__submit MyClassForm__submit Form__submit__edit MyClassForm__submit__edit"
        iconName="check"
        loading={true}
        onClick={[Function]}
        params={Object {}}
        text="Save"
        type="submit"
      />
    </div>
  </fieldset>
</form>
`;

exports[`Form should render the component correctly with validation alert message for each field 1`] = `
<form
  action="edit"
  id="MyClassForm"
  name="MyClassForm"
  onSubmit={[Function]}
>
  <fieldset
    className="Form__form MyClassForm__form"
    disabled={false}
  >
    <div
      className="Form__fields MyClassForm__fields"
    >
      <InputTextWrapper
        className="FormInput MyClassFormInput"
        hidden={false}
        key="name"
        label="Name (internal use)*"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        validationKeys={
          Object {
            "onValidate": undefined,
            "onWarning": undefined,
            "pattern": undefined,
            "required": true,
            "validationAlert": "My Validation Alert Message",
            "warningAlert": "",
          }
        }
        value="My Name"
      />
      <InputTextWrapper
        className="FormInput MyClassFormInput"
        hidden={false}
        key="description"
        label="Description (internal use)"
        name="description"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        validationKeys={
          Object {
            "onValidate": undefined,
            "onWarning": undefined,
            "pattern": undefined,
            "required": undefined,
            "validationAlert": "My Validation Alert Message",
            "warningAlert": "",
          }
        }
        value="My Description"
      />
    </div>
    <div
      className="Form__formActions MyClassForm__formActions"
    >
      <ButtonWrapper
        className="Form__cancel MyClassForm__cancel"
        disabled={false}
        iconName="remove"
        onClick={[Function]}
        params={Object {}}
        text="Cancel"
      />
      <ButtonWrapper
        className="Form__submit MyClassForm__submit Form__submit__edit MyClassForm__submit__edit"
        iconName="check"
        loading={false}
        onClick={[Function]}
        params={Object {}}
        text="Save"
        type="submit"
      />
    </div>
  </fieldset>
</form>
`;

exports[`Form should render the component correctly with warning message for each field 1`] = `
<form
  action="edit"
  id="MyClassForm"
  name="MyClassForm"
  onSubmit={[Function]}
>
  <fieldset
    className="Form__form MyClassForm__form"
    disabled={false}
  >
    <div
      className="Form__fields MyClassForm__fields"
    >
      <InputTextWrapper
        className="FormInput MyClassFormInput"
        hidden={false}
        key="name"
        label="Name (internal use)*"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        validationKeys={
          Object {
            "onValidate": undefined,
            "onWarning": undefined,
            "pattern": undefined,
            "required": true,
            "validationAlert": "",
            "warningAlert": "My Warning Message",
          }
        }
        value="My Name"
      />
      <InputTextWrapper
        className="FormInput MyClassFormInput"
        hidden={false}
        key="description"
        label="Description (internal use)"
        name="description"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        validationKeys={
          Object {
            "onValidate": undefined,
            "onWarning": undefined,
            "pattern": undefined,
            "required": undefined,
            "validationAlert": "",
            "warningAlert": "My Warning Message",
          }
        }
        value="My Description"
      />
    </div>
    <div
      className="Form__formActions MyClassForm__formActions"
    >
      <ButtonWrapper
        className="Form__cancel MyClassForm__cancel"
        disabled={false}
        iconName="remove"
        onClick={[Function]}
        params={Object {}}
        text="Cancel"
      />
      <ButtonWrapper
        className="Form__submit MyClassForm__submit Form__submit__edit MyClassForm__submit__edit"
        iconName="check"
        loading={false}
        onClick={[Function]}
        params={Object {}}
        text="Save"
        type="submit"
      />
    </div>
  </fieldset>
</form>
`;
